Packet++/header/SomeIpLayer.h

```diff
@@ -0,0 +1,361 @@
+	 * @param[in] dataLen Size of the data in bytes
+	 * @param[in] prevLayer A pointer to the previous layer
+	 * @param[in] packet A pointer to the Packet instance where layer will be stored in
+	 */
+	SomeIpLayer(uint8_t *data, size_t dataLen, Layer *prevLayer, Packet *packet)
+		: Layer(data, dataLen, prevLayer, packet)
+	{
+		m_Protocol = SomeIP;
+	}
+
+	/**
+	 * Construct a new layer object
+	 * @param[in] serviceID Service ID
+	 * @param[in] methodID Method ID
+	 * @param[in] clientID Client ID
+	 * @param[in] sessionID Session ID
+	 * @param[in] interfaceVersion Interface Version
+	 * @param[in] type Type of the message
+	 * @param[in] returnCode Return Code
+	 * @param[in] data A pointer to the raw data
+	 * @param[in] dataLen Size of the data in bytes
+	 * holds the reference to a data buffer. This option can be used to reduce the number of copies to generate packets.
+	 */
+	SomeIpLayer(uint16_t serviceID, uint16_t methodID, uint16_t clientID, uint16_t sessionID, uint8_t interfaceVersion,
+				MsgType type, uint8_t returnCode, const uint8_t *const data = nullptr, size_t dataLen = 0);
+
+	/**
+	 * Destroy the layer object
+	 */
+	~SomeIpLayer() {}
+
+	/**
+	 * Get a pointer to the basic SOME/IP header. Notice this points directly to the data, so every change will change
+	 * the actual packet data
+	 * @return A pointer to the @ref someiphdr
+	 */
+	someiphdr *getSomeIpHeader() const { return (someiphdr *)m_Data; }
+
+	/**
+	 * Checks if given port is a SOME/IP protocol port (only Service Discovery ports are checked for now )
+	 * @param[in] port Port to check
+	 * @return true if SOME/IP protocol port, false if not
+	 */
+	static bool isSomeIpPort(uint16_t port);
+
+	/**
+	 * Adds given port to ports that are obtained dynamically by SOME/IP Service discovery.
+	 * These are checked if a new packet is parsed and determine if packet is a SOME/IP packet
+	 * @param[in] port Port to add
+	 */
+	static void addSomeIpPort(uint16_t port);
+
+	/**
+	 * A static method that validates the input data
+	 * @param[in] data The pointer to the beginning of a byte stream of a SOME/IP packet
+	 * @param[in] dataLen The length of the byte stream
+	 * @return true if SOME/IP protocol, false otherwise
+	 */
+	static bool isDataValid(uint8_t *data, uint32_t dataLen);
+
+	/**
+	 * @brief Checks that the SomeIp header signals that the following layer is a SomeIpSd layer.
+	 * @param[in] serviceId ServiceId of the SomeIp layer
+	 * @param[in] methodId MethodId of the SomeIp layer
+	 * @return true If the packet contains a SomeIpSd layer, false ottherwise
+	 */
+	static bool isSomeIpSdPacket(uint16_t serviceId, uint16_t methodId);
+
+	/**
+	 * @brief Checks whether the packet contains a SomeIpTp layer by checking the message type for the TP flag
+	 * @param[in] msgType MsgType that has to be checked
+	 * @return true if packet contains a SomeIpTp layer, false if otherwise
+	 */
+	static bool isSomeIpTpPacket(uint8_t msgType);

```

Comment:
I'm not sure these methods should be public. For now it's ok because we don't yet have SD and TP, but once we have them the type of the layer would be `SomeIpSdLayer` or `SomeIpTpLayer`, so those methods might be confusing. What do you think?